### ISSUE TITLE
[Feature Request] Dynamic batch size for gradient accumulation

### ISSUE THREAD

**Issue by andersonbcdefg** [huggingface/accelerate#1360]
Currently, the `Accelerator` is initialized with a fixed number of gradient accumulation steps. This means you will have the same batch size throughout training. However, many practitioners advocate for various forms of dynamic effective batch sizes, whether it's a [batch size ramp-up at the beginning of training](https://arxiv.org/pdf/2212.14034.pdf), [adaptive batch sizes](https://arxiv.org/abs/1610.05792), or [increasing the batch size at the end of training instead of decaying the learning rate](https://arxiv.org/abs/1711.00489).

Right now as far as I can tell, the options are either a static batch size; or manually handling gradient accumulation (the first, not-recommended way in the [gradient accumulation tutorial](https://huggingface.co/docs/accelerate/v0.18.0/en/usage_guides/gradient_accumulation)). The latter seems likely to kill performance on multi-GPU setups because of synchronization, though it's working fine for me on a single GPU. Would appreciate a feature to allow me to dynamically increase the number of accumulation iterations throughout training!

**Comment by sgugger** [comment]
You can already do this by doing the gradient accumulation manually. You just need to use the `no_sync` context manager for the backward pass when you don't want to do an optimizer step.

**Comment by muellerzr** [comment]
The other option would be to modify `accelerator.gradient_accumulation_steps` manually, however do note this should be done *before* iterating the training dataloader if so

**Comment by andersonbcdefg** [comment]
Okay, so would this be the right way to do it using the strategy suggested by Zachary?

```
for index, batch in enumerate(train_dataloader):
    with accelerator.accumulate():
        input_ids, labels = batch["input_ids"], batch["targets"]
        micro_batch_loss = model(input_ids=input_ids, labels=labels).loss
        accelerator.backward(loss)
        if accelerator.sync_gradients:
            accelerator.clip_grad_norm_(model.parameters(), 0.5)
        optimizer.step()
        scheduler.step()
        optimizer.zero_grad(set_to_none=True)
        wandb.log({
            "micro_batch_loss": micro_batch_loss,
            "lr": scheduler.get_last_lr()[0],
        })
        
        # trying this--if we are done accumulating, linearly increase the effective batch size
        if accelerator.sync_gradients and accelerator.gradient_accumulation_steps < max_batch_size // micro_batch_size:
            accelerator.gradient_accumulation_steps += 1
```
